app/models/manageiq/providers/azure/cloud_manager/event_catcher/stream.rb

```diff
@@ -1,67 +1,112 @@
   #
   def initialize(ems)
     @ems = ems
     @collecting_events = false
     @since = nil
+    @connection = nil
   end
 
-  # Start capturing events
+  # Sets a boolean used by the +each_batch+ method that indicates
+  # that events should start/keep being captured.
+  #
   def start
     @collecting_events = true
   end
 
-  # Stop capturing events
+  # Sets a boolean used by the +each_batch+ method that indicates
+  # that events should stop being captured.
+  #
   def stop
     @collecting_events = false
   end
 
+  # Used internally by the Runner#monitor_events method.
+  #
   def each_batch
     while @collecting_events
-      yield get_events.collect { |e| JSON.parse(e) }
+      yield get_events.collect(&:to_hash)
     end
   end
 
   private
 
+  # Get a list of events that have happened since the most recent event time.
+  #
+  # Because Azure event timestamps are not necessarily stamped in order, an
+  # issue occurs where we could accidentally skip over events that happen
+  # in quick succession. We must therefore begin our query a couple minutes
+  # back from our most recent timestamp, and filter out any duplicates.
+  #
+  # See https://bugzilla.redhat.com/show_bug.cgi?id=1724312 for details.
+  #
   def get_events
-    # Grab only events for the last minute if this is the first poll
-    filter = @since ? "eventTimestamp ge #{@since}" : "eventTimestamp ge #{startup_interval}"
-    fields = 'authorization,description,eventName,eventTimestamp,resourceGroupName,resourceProviderName,resourceId,resourceType'
-    events = connection.list(:filter => filter, :select => fields, :all => true).sort_by(&:event_timestamp)
-
-    # HACK: the Azure Insights API does not support the 'gt' (greater than relational operator)
-    # therefore we have to poll from 1 millisecond past the timestamp of the last event to avoid
-    # gathering the same event more than once.
-    @since = one_ms_from_last_timestamp(events) unless events.empty?
-    events
+    filter = "eventTimestamp ge #{most_recent_time}"
+    connection.list(:filter => filter, :select => SELECT_FIELDS, :all => true)
   end
 
+  # When the appliance first starts, or is restarted, start looking for events
+  # from a fixed, recent point in the past.
+  #
   def startup_interval
-    format_timestamp(1.minute.ago)
+    format_timestamp(2.minutes.ago)
   end
 
-  def one_ms_from_last_timestamp(events)
-    time = Time.at(one_ms_from_last_timestamp_as_f(events)).utc
-    format_timestamp(time)
-  end
+  # Retrieve the most recent Azure event minus 2 minutes, or the startup interval
+  # if no records are found.
+  #
+  # Go back a maximum of 1 hour if the newest record is older than that to avoid
+  # hitting the account's request limits. This should only happen in practice if
+  # the appliance has been down for a while and was restarted.
+  #
+  def most_recent_time
+    result = EventStream.where(:source => 'AZURE', :ems_id => ems.id).maximum(:timestamp)
```

Comment:
I think you missed the point of https://github.com/ManageIQ/manageiq-providers-azure/pull/350#discussion_r326146682

Reply:
Well, I also wanted to test it quick. :)

Reply:
btw, you can simply `ems: ems` instead of `ems_id: ems.id`, and you don't need a `source` filter with `ems`.
or even
`ems.ems_events.maximum(:timestamp)`

Reply:
@AlexanderZagaynov It would actually have to be `:ext_management_system => ems`. :)

But, anyway, I ditched it.